### ISSUE TITLE
having print statements in dataset iterator makes it much slower

### ISSUE THREAD

**Issue by ramsane** [tensorflow/tensorflow#35873]
I was trying to write the model training by using gradient Tape. Somehow, it is slower than Keras.fit() method. On debugging lot of time. I came to know that the reason was having some other code while iterating through the dataset, I was storing few things in a list and a print statements.

It is what slowing the program. here is the code that is required. I've run it multiple times.
![image](https://user-images.githubusercontent.com/14745146/72365972-55ba2d80-371f-11ea-9524-757d4fb9197d.png)

And after adding a print statement, 
![image](https://user-images.githubusercontent.com/14745146/72366050-7b473700-371f-11ea-88d5-ba93b33d5da5.png)

The time got reduced if I reduce the print statements. 
![image](https://user-images.githubusercontent.com/14745146/72366092-974ad880-371f-11ea-9822-6f84c3190cc4.png)

Can anyone explain this?

**Comment by ramsane** [comment]
here is the code that I've used for train_step..

`@tf.function
def train_step__(model, batch, labels, loss_fn, optimizer, acc_fn):
    with tf.GradientTape() as tape:
        # recored all the values on a tape in forward pass
        out = model(batch)
        loss = loss_fn(y_true=labels, y_pred=out)
        acc = acc_fn(y_true=labels, y_pred=out)
    # backward pass : grad of parameters w.r.t Loss
    gradients = tape.gradient(loss, model.trainable_variables)
    optimizer.apply_gradients(grads_and_vars=zip(gradients, model.trainable_variables))
    # return the loss before applying the gradients
    return loss, acc`

**Comment by mihaimaruseac** [comment]
Not a TF issue:

```python
import timeit
import sys

def no_print(n=100):
  s = 0
  for x in range(n):
    s += x

def do_print(n=100):
  s = 0
  for x in range(n):
    s += x
    print(s)


n=10
print(timeit.timeit(stmt=lambda:no_print(n)), file=sys.stderr)
print(timeit.timeit(stmt=lambda:do_print(n)), file=sys.stderr
```

```
(gh35873) mihaimaruseac@ankh:/tmp/gh35873$ python test.py >/dev/null
0.6009224290028214
4.1080495161004364
```

**Comment by ramsane** [comment]
Ohkkk. Thanks !! 

I will look into the tf.summary.

Thank you once again.
